### PR TITLE
Llama2 contextmodel

### PR DESCRIPTION
**Pull request by riensou**
added capability to use other transformer architectures easily, including Llama, as discussed in #6

### CODE REVIEW

File: conf/models.yml
Comment (nelson-lojo) [inline conf/models.yml:1]: I think we should keep the number of args here at a minimum -- i.e. just specify "llama2" or "gpt2" and infer "transformer" later

File: conf/train/train_mlp.yml [hunk 1]
```diff
@@ -22,5 +22,5 @@ x_dist:
 steps: 500000
 log_freq: 100
 checkpoint_freq: 5000
-model: *std_gpt2
+model: *std_llama2
```
Comment (nelson-lojo) [inline conf/train/train_mlp.yml:25]: We should make a separate config file instead of supplanting. That'll let us verify that our code has "functional-"parity with the Garg et al. repo.

File: src/models/__init__.py [hunk 2]
```diff
@@ -11,7 +11,7 @@
 from .decision_tree import DecisionTreeModel
 
 MODELS: dict[str, type[ContextModel]] = {
-    "gpt2" : TransformerModel,
+    "transformer" : TransformerModel,
```
Comment (nelson-lojo) [inline src/models/__init__.py:14]: see comment on `conf/models.yml`

File: src/models/transformer.py [hunk 3]
```diff
@@ -1,29 +1,37 @@
...
+from transformers import GPT2Config, GPT2Model, LlamaConfig, LlamaModel
 from torch import nn
 
 from core import ContextModel
 
-
+ARCHITECTURES = {
+    "gpt2": {'config': GPT2Config, 'model': GPT2Model},
+    "llama2": {'config': LlamaConfig, 'model': LlamaModel}
+}
```
Comment (nelson-lojo) [inline src/models/transformer.py:10]: See comment on `conf/models.yml`

File: src/models/transformer.py [hunk 4]
```diff
@@ -1,29 +1,37 @@
...
             use_cache=False,
         )
-        self.name = f"gpt2_embd={n_embd}_layer={n_layer}_head={n_head}"
+        self.name = f"{kwargs['architecture']}_embd={n_embd}_layer={n_layer}_head={n_head}"
 
         self.context_length = n_positions
         self._n_dims = x_dim
         self._read_in = nn.Linear(x_dim, n_embd)
-        self._backbone = GPT2Model(configuration)
+        self._backbone = architecture['model'](configuration)
```
Comment (nelson-lojo) [inline src/models/transformer.py:34]: I like the code reuse here, could we take a more delegation-esque approach? For example, have a base class that takes in just the backbone method and have two very thin classes that serve for being either GPT2 or Llama2

**Review by nelson-lojo — COMMENTED** [review]
Looks mostly fine, although I think there's a more scalable way to support multiple transformer architectures. In particular, this approach requires changing code that may have been previously tested and verified, making it easy to break when adding a new architecture (say trajectory transformer).

File: environment.yaml [hunk 5]
```diff
@@ -15,7 +15,7 @@ dependencies:
     - scikit-learn==1.0.2
     # - seaborn==0.11.2
     - tqdm==4.64.0
-    - transformers==4.17.0
+    - transformers==4.39.1
```
Comment (KesavaViswanadha) [inline environment.yaml:18]: Should be 4.39.3

**Comment by KesavaViswanadha** [comment]
I'm wondering how we can integrate this with my own repo's architectures code. I have a working implementation that allows me to construct all sorts of hybrid architectures using mamba blocks, attention, MLPs, etc. and run gradient steps, but optically the code is inelegant and looks hideous. 

My repo: https://github.com/KesavaViswanadha/in-context-updated/

File: conf/models.yml
Comment (nelson-lojo) [reply conf/models.yml:1]: implemented

File: conf/train/train_mlp.yml (hunk 1 above)
Comment (nelson-lojo) [reply conf/train/train_mlp.yml]: implemented

File: src/models/transformer.py (hunk 4 above)
Comment (nelson-lojo) [reply src/models/transformer.py]: implemented

**Review by nelson-lojo — DISMISSED** [review]
lgtm!

**Review by nelson-lojo — APPROVED** [review]
lgtm
